### PR TITLE
[mlir] [arith] add shl overflow flag in Arith and lower to SPIR-V and LLVMIR

### PR DESCRIPTION
**Pull request by yiwu0b11**
There is no `SHL` used in canonicalization in `arith`

### CODE REVIEW

**Review by Hardcode84 — APPROVED** [review]
Nice, thanks

File: mlir/include/mlir/Dialect/Arith/IR/ArithOps.td [hunk 1]
```diff
@@ -791,11 +791,17 @@ def Arith_ShLIOp : Arith_TotalIntBinaryOp<"shli"> {
...
+    "No Unsigned Wrap" and "No Signed Wrap", respectively. If the `nuw` and/or
+    `nsw` flags are present, and an unsigned/signed overflow occurs
+    (respectively), the result is poison.
+
     Example:
 
     ```mlir
     %1 = arith.constant 5 : i8                 // %1 is 0b00000101
     %2 = arith.constant 3 : i8
+    %a = arith.shli %1, %2 overflow<nsw, nuw> : (i8, i8) -> i8
```
Comment (kuhar) [inline mlir/include/mlir/Dialect/Arith/IR/ArithOps.td:804]: nit: Why use the functional type in this example?
```suggestion
    %a = arith.shli %1, %2 overflow<nsw, nuw> : i8  
```

**Review by kuhar — APPROVED** [review]

File: mlir/include/mlir/Dialect/Arith/IR/ArithOps.td (hunk 1 above)
Comment (yi-wu-arm) [reply mlir/include/mlir/Dialect/Arith/IR/ArithOps.td]: (leftover from copy pasting) Thanks for pointing this out, I should've pay more attention to this!

File: mlir/include/mlir/Dialect/Arith/IR/ArithOps.td [hunk 2]
```diff
@@ -791,11 +791,17 @@ def Arith_ShLIOp : Arith_TotalIntBinaryOp<"shli"> {
...
+    `nsw` flags are present, and an unsigned/signed overflow occurs
+    (respectively), the result is poison.
+
     Example:
 
     ```mlir
     %1 = arith.constant 5 : i8                 // %1 is 0b00000101
     %2 = arith.constant 3 : i8
+    %a = arith.shli %1, %2 overflow<nsw, nuw> : i8  
     %3 = arith.shli %1, %2 : (i8, i8) -> i8    // %3 is 0b00101000
```
Comment (gysit) [inline mlir/include/mlir/Dialect/Arith/IR/ArithOps.td:805]:
```suggestion
    %1 = arith.constant 5 : i8  // %1 is 0b00000101
    %2 = arith.constant 3 : i8
    %3 = arith.shli %1, %2 : i8 // %3 is 0b00101000
    %4 = arith.shli %1, %2 overflow<nsw, nuw> : i8  
```
nit: looks like the type on the old example was broken?

**Review by gysit — APPROVED** [review]
LGTM modulo nit!

File: mlir/include/mlir/Dialect/Arith/IR/ArithOps.td (hunk 2 above)
Comment (yi-wu-arm) [reply mlir/include/mlir/Dialect/Arith/IR/ArithOps.td]: Thanks for the review! Done.

**Pull request merged**